docs/USAGE-GUIDE.md

```diff
@@ -1777,6 +1777,216 @@ handle. The handle is considered invalid after **s2n_connection_free** is used.
+
+Use the following APIs to configure and use external pre-shared keys.
+
+```c
+struct s2n_psk* s2n_external_psk_new();
+int s2n_psk_free(struct s2n_psk **psk);
+int s2n_psk_set_identity(struct s2n_psk *psk, const uint8_t *identity, uint16_t identity_size);
+int s2n_psk_set_secret(struct s2n_psk *psk, const uint8_t *secret, uint16_t secret_size);
+int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac);
+int s2n_connection_append_psk(struct s2n_connection *conn, struct s2n_psk *psk);
+int s2n_config_set_psk_mode(struct s2n_config *config, s2n_psk_mode mode);
+int s2n_connection_set_psk_mode(struct s2n_connection *conn, s2n_psk_mode mode);
+```
+
+To create an external PSK a caller must call the **s2n_external_psk_new** API. **s2n_external_psk_new** creates a new external PSK object
+with **S2N_PSK_HMAC_SHA256** as the default PSK hash algorithm. Use **s2n_psk_free** to free the memory allocated to the external
+PSK object created by this API.
+
+**s2n_psk_set_identity** sets the PSK identity for a given PSK. The PSK identity is a unique identifier for the pre-shared secret. The PSK identity is transmitted over the network unencrypted and is a non-secret value. Do not include any confidential information in the identity. Note that the identity value is copied into s2n-tls memory and the user is responsible for freeing the memory associated with the **identity** input.
+
+**s2n_psk_set_secret** sets the secret value for a given PSK. Note that the secret value is copied into s2n-tls memory and the user is responsible for freeing the memory associated with the **secret** input.
+
+**s2n_psk_set_hmac** sets the PSK hash algorithm for a given PSK.
+The supported PSK hash algorithms are listed in the **s2n_psk_hmac** enum. This API overrides the default PSK hash algorithm value of **S2N_PSK_HMAC_SHA256**. Use this API if you wish to select a different cipher suite, as the selected cipher suite must have the same hash algorithm as the PSK.
+
+After adding an identity and secret to the PSK object, append the PSK to the connection using **s2n_connection_append_psk**. This API sets up the connection to be later used by the server PSK extension to select a PSK for the connection. If a PSK with a duplicate identity is found, an error is returned and the PSK is not added to the list. If the connection already contains a list of PSKs **s2n_connection_append_psk** will verify that the PSK mode being set matches that of the list. If the PSK modes differs, an error is returned. If there is no existing list of PSKs present on the connection (the input PSK is the first PSK to be added to the list) the PSK mode of the list is set to the input PSK's mode. The PSK object is copied to the connection and the user is responsible for freeing the memory associated with the PSK object. Note that both the server and client should order the PSKs appended in preference order ranging from highest preferred PSK to lowest preferred PSK.
```

Comment:
```suggestion
Both clients and servers should append their configured PSKs to the connection by calling **s2n_connection_append_psk**. Append the PSKs in order from most preferred to least preferred.
```
I actually think most of the information in this paragraph is irrelevant to the caller. We don't need to describe each error this function might hit.

Reply:
I disagree, this information is necessary to the caller as the **s2n_connection_append_psk** has edge cases. It is important for the customer to know which PSK mode is set on the connection and why they receive an error when the PSK mode does not match with the existing PSK mode for the list. It is a special case of this API that will bite the customer if they are not aware of it.